gpcontrib/arenadata_toolkit/arenadata_toolkit.c

```diff
@@ -1,4 +1,169 @@
+#include <dirent.h>
+#include <sys/stat.h>
+#include <ctype.h>
+
 #include "postgres.h"
 #include "fmgr.h"
+#include "funcapi.h"
+#include "miscadmin.h"
+#include "storage/fd.h"
+#include "nodes/execnodes.h"
+#include "cdb/cdbvars.h"
+#include "utils/builtins.h"
+#include "utils/relfilenodemap.h"
+#include "utils/timestamp.h"
+#include "catalog/pg_tablespace.h"
 
 PG_MODULE_MAGIC;
+
+typedef struct
+{
+	char     *datpath;
+	DIR      *dirdesc;
+	TupleDesc tupdesc;
+} user_fctx_data;
+
+/*
+ * Name of file must be "XXX.X" or "XXX"
+ * where X is digit. Count of digits is not set.
+ */
+static bool should_skip_file(const char *filename)
+{
+	int pos = 0;
+	int filenamelen = strlen(filename);
+	bool found_dot = false;
+
+	for (; pos < filenamelen; pos++)
+	{
+		if (isdigit(filename[pos]))
+			continue;
+
+		if ('.' == filename[pos] && !found_dot && pos != 0)
+		{
+			found_dot = true;
+			continue;
+		}
+
+		return true;
+	}
+
+	return false;
+}
+
+PG_FUNCTION_INFO_V1(adb_get_relfilenodes);
+Datum adb_get_relfilenodes(PG_FUNCTION_ARGS)
+{
+	Oid              datoid = MyDatabaseId;
+	Oid              tablespace_oid = PG_GETARG_OID(0);
+
+	struct dirent   *direntry;
+	user_fctx_data  *fctx_data;
+	FuncCallContext *funcctx;
+
+	if (tablespace_oid == GLOBALTABLESPACE_OID)
+		datoid = 0;
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		MemoryContext oldcontext;
+
+		funcctx = SRF_FIRSTCALL_INIT();
+		oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+
+		fctx_data = palloc(sizeof(user_fctx_data));
+		fctx_data->datpath = GetDatabasePath(datoid, tablespace_oid);
+		fctx_data->dirdesc = AllocateDir(fctx_data->datpath);
+
+		if (!fctx_data->dirdesc)
+		{
+			/* Nothing to do: empty tablespace (maybe it has been just created)*/
+			MemoryContextSwitchTo(oldcontext);
+			SRF_RETURN_DONE(funcctx);
+		}
+
+		if (get_call_result_type(fcinfo, NULL, &fctx_data->tupdesc)
+				!= TYPEFUNC_COMPOSITE)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("return type must be a row type")));
+
+		funcctx->attinmeta = TupleDescGetAttInMetadata(fctx_data->tupdesc);
+		funcctx->user_fctx = fctx_data;
+
+		MemoryContextSwitchTo(oldcontext);
+	}
+
+	funcctx = SRF_PERCALL_SETUP();
+	fctx_data = (user_fctx_data *) funcctx->user_fctx;
+
+	while ((direntry = ReadDir(fctx_data->dirdesc, fctx_data->datpath)) != NULL)
+	{
+		struct stat fst;
+		Datum       values[10];
+		bool        nulls[10];
+		char       *relfilenode, *ptr, *filename;
+		Oid         reloid;
+		Oid         relfilenode_oid;
+		HeapTuple   tuple;
+
+		CHECK_FOR_INTERRUPTS();
+
+		if (direntry->d_type == DT_DIR)
+			continue;
+
+		if (should_skip_file(direntry->d_name))
```

Comment:
There is a standard check: https://github.com/arenadata/gpdb/blob/50cc6aa615b6d2e13417040827f3e6164f43a984/src/backend/utils/adt/regproc.c#L932-L940

Reply:
In general, I think that this does not correspond to the original specification:
1) don't provide any information for files starting with pg, t_, . and .., as well as ending with _fsm, _vm and _init
2) if there is a dot in the file name, then for the point 3) take up to the dot
3) try to find the relation corresponding to the file name
4) if successful - show it, otherwise show NULL